src/jupyterlabcontroller/services/prepuller.py

```diff
@@ -0,0 +1,883 @@
+        """We have an annotated image with many tags.  We want to work
+        through these tags and return an image with a canonical pull tag
+        and a single, but possibly compound, (e.g.
+        "Recommended (Weekly 2022_44)") display name.
+        """
+
+        recommended = self.config.recommended_tag
+        primary_tag = ""
+        primary_name = ""
+        other_names: List[str] = list()
+        best_tag_type: Optional[RSPTagType] = None
+        best_nonalias_tag_type: Optional[RSPTagType] = None
+
+        if recommended in img.tags:
+            primary_tag = recommended
+            primary_name = img.tags[recommended]
+            img.all_tags.append(recommended)
+
+        img.tagobjs = RSPTagList()
+        img.tagobjs.all_tags = list()
+        for t in img.tags:
+            # Turn each text tag into a stripped-down RSPTag...
+            img.tagobjs.all_tags.append(
+                RSPTag.from_tag(
+                    t, digest=img.digest, alias_tags=img.known_alias_tags
+                )
+            )
+        # And then sort them.
+        img.tagobjs.sort_all_tags()
+
+        # Now that they're sorted, construct the return image with
+        # a compound display_name, a canonical tag, and a favored
+        # image type.
+        for t_obj in img.tagobjs.all_tags:
+            if primary_name == "":
+                primary_name = t_obj.display_name
+                primary_tag = t_obj.tag
+            else:
+                other_names.append(t_obj.display_name)
+            if best_tag_type is None:
+                best_tag_type = t_obj.image_type
+            if (
+                best_nonalias_tag_type is None
+                and t_obj.image_type != RSPTagType.ALIAS
+            ):
+                best_nonalias_tag_type = t_obj.image_type
+        tag_description: str = primary_name
+        if other_names:
+            tag_description += f" ({', '.join(x for x in other_names)})"
+        # The "best" tag is the one corresponding to the highest-sorted
+        # tag.  all_tags will be sorted according to the tagobjs, which is
+        # to say, according to type.
+        # And we know the digest, so, hey, let's make that part of the pull
+        # path!
+        untagged = img.path.split(":")[0]
+        path = f"{untagged}:{primary_tag}@{img.digest}"
+        return NodeTagImage(
+            path=path,
+            name=tag_description,
+            digest=img.digest,
+            tags=copy(img.tags),
+            size=img.size,
+            prepulled=img.prepulled,
+            best_tag=primary_tag,
+            all_tags=[t.tag for t in img.tagobjs.all_tags],
+            nodes=copy(img.nodes),
+            known_alias_tags=copy(img.known_alias_tags),
+            tagobjs=copy(img.tagobjs),
+            best_tag_type=best_tag_type,
+            best_nonalias_tag_type=best_nonalias_tag_type,
+        )
+
+    async def get_menu_images(self) -> DisplayImages:
```

Comment:
Yeah, this is a FIXME later.  I need to rewrite the Tag class to fit our needs better (right now it's basically straight outta cachemachine).  But honestly it's lower-priority, since it's all in-memory stuff, and even a couple thousand tags is at most going to represent a handful of megabytes.  The execution time in this thing is going to be mostly in the docker API calls and some in the k8s API calls.  That's almost certainly going to dwarf any internal calculation time.